notte/env.py

```diff
@@ -71,46 +71,68 @@ def previous_actions(self) -> list[Action] | None:
         if len(self._trajectory) <= 1:
             return None
         previous_obs: Observation = self._trajectory[-2]
-        if isinstance(previous_obs, PreObservation):
+        if not previous_obs.has_space():
             return None  # we don't have a space for pre-observations
-        if self.context.snapshot.clean_url != previous_obs.clean_url:
+        # if self.context.snapshot.clean_url != previous_obs.clean_url:
+        if self.obs.clean_url != previous_obs.clean_url:
             return None  # the page has significantly changed
         return previous_obs.space.actions(status="all")
 
+    @property
+    def obs(self) -> Observation:
+        if len(self._trajectory) <= 0:
+            raise ValueError("Need to observe first to get a context.")
+        return self._trajectory[-1]
+
     # ---------------------------- observe, step functions ----------------------------
 
-    def _preobserve(self, snapshot: BrowserSnapshot) -> PreObservation:
+    def _preobserve(self, snapshot: BrowserSnapshot) -> Observation:
         self._context = BrowserSnapshotToContextPipe.forward(snapshot)
-        preobs = PreObservation(_url=snapshot.url, _screenshot=snapshot.screenshot, _space=None)
+        preobs = Observation(url=snapshot.url, screenshot=snapshot.screenshot)
         self._trajectory.append(preobs)
         return preobs
 
-    def _obslisting(self, preobs: PreObservation) -> Observation:
-        space = self._context_to_action_space_pipe.forward(self.context, self.previous_actions)
-        obs = Observation(_url=preobs.url, _screenshot=preobs.screenshot, _space=space)
-        self._trajectory[-1] = obs  # update the last observation with the new space
-        return obs
+    async def _obslisting(self) -> Observation:
+        space = await self._context_to_action_space_pipe.forward_async(self.context, self.previous_actions)
+        async with self._obs_lock:
+            self.obs.space = space
+            return self.obs
+
+    async def _obsscraping(self) -> Observation:
+        # TODO: move this to `scrape` if this is not used more than once
+        data = await self._data_scraping_pipe.forward_async(self.context)
+        async with self._obs_lock:
+            self.obs.data = data
+            return self.obs
+
+    async def _obsfull(self) -> Observation:
+        _ = await asyncio.gather(
```

Comment:
Hmmm OK I understand. But I am really NOT a fan of this behavior. Isn't there something thread-safer? And less far-fetched